### PR TITLE
bug(akkabootstrapper): Retry fetching seeds endpoint before resorting…

### PR DESCRIPTION
**Pull request by vishramachandran**
… to creating new cluster

**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Testing indicates that seeds endpoint retrieval can fail because of load balancer routing request to a node going down during a rolling upgrade.

This fix puts in place retry logic to ensure that we really solidify the observation of no cluster before creating one.

In addition, a rolling upgrade should not bring a node down while a new node is being brought up. Node downing and upping operations should be disjoint and carried out at different times.

### CODE REVIEW

File: akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala [hunk 1]
```diff
@@ -24,40 +24,44 @@ abstract class ClusterSeedDiscovery(val cluster: Cluster,
...
+    do {
+      try {
+        logger.info(s"Trying to fetch seeds from $seedsEndpoint ... $retriesRemaining retries remaining.")
+        response = Http(seedsEndpoint).timeout(2000, 2000).asString
+        logger.info(s"Seeds endpoint returned a ${response.code}. Response body was ${response.body}")
+      } catch {
+        case NonFatal(e) =>
+          logger.error(s"Seeds endpoint $seedsEndpoint call threw an exception", e)
+      }
+      Thread.sleep(settings.seedsHttpSleepBetweenRetries.toMillis)
```
Comment (tjackpaul) [inline akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala]: Isn't it better to add an exponential backoff ?

Comment (vishramachandran) [reply akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala]: This is during bootstrap, initial formation of cluster for the very first time. We should form the soon and not back-off exponentially.

File: akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala [hunk 2]
```diff
@@ -24,40 +24,44 @@ abstract class ClusterSeedDiscovery(val cluster: Cluster,
...
 
   protected def discoverExistingCluster: Seq[Address] = {
 
     val seedsEndpoint = settings.seedsBaseUrl + settings.seedsPath
+    var response: HttpResponse[String] = null
+    var retriesRemaining = settings.seedsHttpRetries
+    do {
+      try {
+        logger.info(s"Trying to fetch seeds from $seedsEndpoint ... $retriesRemaining retries remaining.")
+        response = Http(seedsEndpoint).timeout(2000, 2000).asString
```
Comment (velvia) [inline akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala]: should this timeout be configurable also?

**Comment by vishramachandran** [comment]
@velvia Currently there is a very serious bug/regression in the testing setup because of the _ns change.

We use the FormatConversion.prometheus dataset for shard key configuration which uses "_ns" as shard key, but we ingest and query with "app". 

Lets use one format and reduce confusion.

File: akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala (hunk 2 above)
Comment (vishramachandran) [reply akka-bootstrapper/src/main/scala/filodb/akkabootstrapper/ClusterSeedDiscovery.scala]: Lets leave that out for now.

**Pull request merged**